cleanlab/datalab/factory.py

```diff
@@ -0,0 +1,177 @@
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+"""The factory module provides a factory class for constructing concrete issue managers
+and a decorator for registering new issue managers.
+
+This module provides the :py:meth:`register` decorator for users to register new subclasses of
+:py:class:`IssueManager <cleanlab.datalab.issue_manager.issue_manager.IssueManager>`
+in the registry.
+
+
+Note
+----
+
+The :class:`REGISTRY` variable is used by the factory class to keep track
+of registered issue managers.
+The factory class is used as an implementation detail by
+:py:class:`Datalab <cleanlab.datalab.datalab.Datalab>`,
+which provides a simplified API for constructing concrete issue managers.
+:py:class:`Datalab <cleanlab.datalab.datalab.Datalab>` is intended to be used by users
+and provides detailed documentation on how to use the API.
+
+Warning
+-------
+Neither the :class:`REGISTRY` variable nor the factory class should be used directly by users.
+"""
+from __future__ import annotations
+
+from typing import Dict, List, Type
+
+from cleanlab.datalab.issue_manager import (
+    IssueManager,
+    LabelIssueManager,
+    NearDuplicateIssueManager,
+    OutlierIssueManager,
+    NonIIDIssueManager,
+)
+
+
+REGISTRY: Dict[str, Type[IssueManager]] = {
+    "outlier": OutlierIssueManager,
+    "label": LabelIssueManager,
+    "near_duplicate": NearDuplicateIssueManager,
+    "non_iid": NonIIDIssueManager,
+}
+"""Registry of issue managers that can be constructed from a string 
+and used in the Datalab class.
+
+:meta hide-value:
+
+Currently, the following issue managers are registered by default:
+
+- ``"outlier"``: :py:class:`OutlierIssueManager <cleanlab.datalab.issue_manager.outlier.OutlierIssueManager>`
+- ``"label"``: :py:class:`LabelIssueManager <cleanlab.datalab.issue_manager.label.LabelIssueManager>`
+- ``"near_duplicate"``: :py:class:`NearDuplicateIssueManager <cleanlab.datalab.issue_manager.duplicate.NearDuplicateIssueManager>`
+- ``"non_iid"``: :py:class:`NonIIDIssueManager <cleanlab.datalab.issue_manager.noniid.NonIIDIssueManager>`
+
+Warning
+-------
+This variable should not be used directly by users.
+"""
+
+
+def list_possible_issue_types() -> List[str]:
+    """Returns a list of all registered issue types.
+
+    By default, these will be run by Datalab if possible.
+
+    See Also
+    --------
+    :py:class:`REGISTRY <cleanlab.datalab.factory.REGISTRY>` : The default issue types and their corresponding issue managers can be found here.
+    """
+    return list(REGISTRY.keys())
+
+
+def list_default_issue_types() -> List[str]:
+    """Returns a list of the issue types that are run by default by Datalab.
+
+    See Also
+    --------
+    :py:class:`REGISTRY <cleanlab.datalab.factory.REGISTRY>` : The default issue types and their corresponding issue managers can be found here.
+    """
+    return ["label", "outlier", "near_duplicate"]
```

Comment:
These methods should both be made user-facing? I'd imagine they are simply in the datalab.py file so users can actually find these methods.

Reply:
Also the two methods' docstrings to do not clearly describe what is their difference.
I thought list_possible is listing all built-in issue types, regardless whether they are audited by default or not. And list_default is the subset of these built-in issue types that are audited by default.